src/compiler/sys.ts

```diff
@@ -232,6 +231,83 @@ namespace ts {
             const platform: string = _os.platform();
             const useCaseSensitiveFileNames = isFileSystemCaseSensitive();
 
+            function fsWatchFile(fileName: string, callback: FileWatcherCallback, pollingInterval?: number): FileWatcher {
+                _fs.watchFile(fileName, { persistent: true, interval: pollingInterval || 250 }, fileChanged);
+                return {
+                    close: () => _fs.unwatchFile(fileName, fileChanged)
+                };
+
+                function fileChanged(curr: any, prev: any) {
+                    const isCurrZero = +curr.mtime === 0;
+                    const isPrevZero = +prev.mtime === 0;
+                    const created = !isCurrZero && isPrevZero;
+                    const deleted = isCurrZero && !isPrevZero;
+
+                    const eventKind = created
+                        ? FileWatcherEventKind.Created
+                        : deleted
+                            ? FileWatcherEventKind.Deleted
+                            : FileWatcherEventKind.Changed;
+
+                    if (eventKind === FileWatcherEventKind.Changed && +curr.mtime <= +prev.mtime) {
+                        return;
+                    }
+
+                    callback(fileName, eventKind);
+                }
+            }
+
+            function fsWatchDirectory(directoryName: string, callback: (eventName: string, relativeFileName: string) => void, recursive?: boolean): FileWatcher {
+                // Node 4.0 `fs.watch` function supports the "recursive" option on both OSX and Windows
+                // (ref: https://github.com/nodejs/node/pull/2649 and https://github.com/Microsoft/TypeScript/issues/4643)
+                let options: any;
+                let watcher = !directoryExists(directoryName) ?
+                    watchMissingDirectory() :
+                    watchPresentDirectory();
+                return {
+                    close: () => {
+                        watcher.close();
+                    }
+                };
+
+                function watchPresentDirectory(): FileWatcher {
+                    if (options === undefined) {
+                        if (isNode4OrLater && (process.platform === "win32" || process.platform === "darwin")) {
+                            options = { persistent: true, recursive: !!recursive };

```

Comment:
Does anything bad happen on earlier node versions or linux if you do pass this option?

Reply:
Not sure, but i have kept our initial logic here for now. This can be investigated further later on